import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java

```diff
@@ -0,0 +1,255 @@
+      log.error("sopInstanceUID={}, removeBackup failed.", fileName, ex);
+    }
+  }
+
+  public abstract class UploadAsyncJob implements Runnable {
+    protected BackupState backupState;
+    protected String uniqueFileName;
+    protected int attemptNumber;
+
+    public UploadAsyncJob(BackupState backupState) {
+      this.backupState = backupState;
+      this.uniqueFileName = backupState.getUniqueFileName();
+      this.attemptNumber = attemptsAmount - backupState.getAttemptsCountdown();
+    }
+
+    protected void logUploadFailed(Exception e) {
+      log.error("sopInstanceUID={}, resend attempt № {} - failed.", uniqueFileName, attemptNumber, e);
+    }
+
+    protected void logSuccessUpload() {
+      log.debug("sopInstanceUID={}, resend attempt № {}, - successful.", uniqueFileName, attemptNumber);
+    }
+
+    protected InputStream readBackupExceptionally() throws CompletionException {
+      try {
+        return backupUploader.doReadBackup(uniqueFileName);
+      } catch (BackupException ex) {
+        log.error("sopInstanceUID={}, read backup failed.", uniqueFileName, ex.getCause());
+        throw new CompletionException(ex);
+      }
+    }
+  }
+
+  public class DicomDestinationUploadAsyncJob extends UploadAsyncJob {
+
+    private CStoreSender cStoreSender;
+    private AetDictionary.Aet target;
+    private String sopInstanceUid;
+    private String sopClassUidBackupState;
+
+    public DicomDestinationUploadAsyncJob(
+        CStoreSender cStoreSender,
+        BackupState backupState,
+        AetDictionary.Aet target,
+        String sopInstanceUid,
+        String sopClassUidBackupState) {
+      super(backupState);
+      this.cStoreSender = cStoreSender;
+      this.target = target;
+      this.sopInstanceUid = sopInstanceUid;
+      this.sopClassUidBackupState = sopClassUidBackupState;
+    }
+
+    @Override
+    public void run() {
+      try {
+        InputStream inputStream = readBackupExceptionally();
+        cStoreSender.cstore(target, sopInstanceUid, sopClassUidBackupState, inputStream);
+        logSuccessUpload();
+      } catch (IOException io) {
+        logUploadFailed(io);
+
+        if (backupState.getAttemptsCountdown() > 0) {
+            try {
+              scheduleUploadWithDelay(
+                  backupState,
+                  new DicomDestinationUploadAsyncJob(
+                      cStoreSender,
+                      backupState,
+                      target,
+                      sopInstanceUid,
+                      sopClassUidBackupState),
+                  delayCalculator.getExponentialDelayMillis(
+                      backupState.getAttemptsCountdown(),
+                      attemptsAmount))
+                  .get();
+
+            } catch (BackupException | ExecutionException | InterruptedException ex) {
+              throw new CompletionException(ex);
+            }
+            MonitoringService.addEvent(Event.CSTORE_BACKUP_ERROR);
```

Comment:
I think we could miss logging this error if the try block throws an exception above.
I think we should move this to the beginning of the outer catch block after logUploadFailed(io)

Reply:
Yes, agree. Have checked and refactored adding of  events by schema: one destination send - one event (add event only on last send attempt)